exporter/tests/test_export.py

```diff
@@ -0,0 +1,138 @@
+"""
+Tests for export
+"""
+
+from __future__ import unicode_literals
+from tempfile import mkdtemp
+from shutil import rmtree
+import os
+
+from archive import Archive
+from django.core.files.storage import default_storage
+from django.utils.text import slugify
+
+from learningresources.tests.base import LoreTestCase
+from learningresources.models import (
+    LearningResource,
+    LearningResourceType,
+)
+from learningresources.api import update_description_path
+from exporter.api import (
+    export_resources_to_directory,
+    export_resources_to_tarball,
+)
+from exporter.tasks import export_resources
+from learningresources.api import create_resource
+from importer.tasks import import_file
+
+
+class TestExport(LoreTestCase):
+    """
+    Tests for export
+    """
+
+    def setUp(self):
+        super(TestExport, self).setUp()
+
+        # Add some LearningResources on top of the default to make things
+        # interesting.
+        tarball_file = self.get_course_single_tarball()
+        import_file(
+            tarball_file, self.repo.id, self.user.id)
+
+        # Add a resource with a '/' in the title and too many characters.
+        resource = create_resource(
+            course=self.repo.course_set.first(),
+            resource_type=LearningResourceType.objects.first().name,
+            title="//x"*300,
+            content_xml="",
+            mpath="",
+            url_name=None,
+            parent=None,
+            dpath=''
+        )
+        update_description_path(resource)
+
+    def assert_resource_directory(self, resources, tempdir):
+        """Assert that files are present with correct content."""
+        def make_name(resource):
+            """Format expected filename."""
+            return "{id}_{title}.xml".format(
+                id=resource.id,
+                title=slugify(resource.title)[:200],
+            )
+
+        # Verify that content XML exists in right place.
+        for resource in resources:
+            type_name = resource.learning_resource_type.name
+            abs_path = os.path.join(tempdir, type_name, make_name(resource))
+            self.assertTrue(os.path.exists(abs_path))
+            with open(abs_path) as f:
+                self.assertEqual(f.read(), resource.content_xml)
+
+            # Verify static asset placement.
+            for static_asset in resource.static_assets.all():
+                relative_path = static_asset.asset.name.replace(
+                    "assets/", "static/")
+                abs_path = os.path.join(tempdir, relative_path)
+                self.assertTrue(os.path.exists(abs_path))
+                with open(abs_path, 'rb') as f:
+                    self.assertEqual(f.read(), static_asset.asset.read())
```

Comment:
This optional, since the files in the test suite _should_ always be tiny, but it's technically better to not read in the files all at once. This _could_ cause problems running tests on restricted machines, like the Travis VMs.

Something like this could be used to limit the amount of memory used to read a file:

```python
from hashlib import sha256

def get_hash(filename):
    """
    Get a file's hash without reading it all
    into memory at once.
    """
    read_size = 2**20
    signature = sha256()
    with open(filename, "rb") as handle:
        chunk = handle.read(read_size)
        while len(chunk) > 0:
            signature.update(chunk)
            chunk = handle.read(read_size)
    return signature.hexdigest()
```